File: mlforge_sdk/mlforge_sdk/training.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel

from .http import HttpClient

class TrainRun(BaseModel):
    id: str
    run_number: int
    model_id: str
    model_name: str
    dataset_id: str
    dataset_name: str
    task: str
    status: str
    epochs_done: int
    total_epochs: int
    best_metric: Dict[str, float]
    final_loss: float
    duration: str
    created_at: float
    completed_at: Optional[float] = None
    hyperparams: Dict[str, Any]

class TrainingClient:
    def __init__(self, http: HttpClient):
        self._http = http

    def list_runs(self) -> List[TrainRun]:
        data = self._http.get("/train/runs")
        return [TrainRun(**r) for r in data]

    def start(self, model_id: str, dataset_id: str, task: str, 
              params: Dict[str, Any], project_id: str = "default") -> Dict[str, Any]:
        payload = {
            "model_id": model_id,
            "dataset_id": dataset_id,
            "task": task,
            "hyperparams": params,
            "augmentation": {},
            "scheduler": {},
            "project_id": project_id
        }
        return self._http.post("/train/start", json_body=payload)

    def get_history(self, run_id: str) -> List[Dict[str, Any]]:
        data = self._http.get(f"/train/runs/{run_id}/history")
        if data is None:
            return []
        return data

File: packages/mlforge_js_sdk/src/training.ts
import type { HttpClient } from "./http.js";

export interface TrainRun {
  id: string;
  model_id: string;
  dataset_id: string;
  task: string;
  status: "queued" | "running" | "completed" | "failed" | "stopped";
  epoch: number;
  total_epochs: number;
  metrics: Record<string, number>;
  created_at: string;
  completed_at?: string;
}

export interface TrainStartOptions {
  model_id: string;
  dataset_id: string;
  task: string;
  epochs?: number;
  batch_size?: number;
  imgsz?: number;
  lr?: number;
  optimizer?: string;
  device?: string;
  augmentation?: Record<string, unknown>;
  hyperparams?: Record<string, unknown>;
}

export class TrainingClient {
  constructor(private http: HttpClient) {}

  async start(opts: TrainStartOptions): Promise<{ run_id: string }> {
    return this.http.post<{ run_id: string }>("/train/start", opts);
  }

  async stop(runId: string): Promise<void> {
    return this.http.post<void>(`/train/runs/${runId}/stop`);
  }

  async pause(runId: string): Promise<void> {
    return this.http.post<void>(`/train/runs/${runId}/pause`);
  }

  async resume(runId: string): Promise<void> {
    return this.http.post<void>(`/train/runs/${runId}/resume`);
  }

  async get(runId: string): Promise<TrainRun> {
    return this.http.get<TrainRun>(`/train/runs/${runId}`);
  }

  async list(): Promise<TrainRun[]> {
    return this.http.get<TrainRun[]>("/train/runs");
  }

  async getHistory(runId: string): Promise<{ epoch: number; metrics: Record<string, number> }[]> {
    return this.http.get<{ epoch: number; metrics: Record<string, number> }[]>(
      `/train/runs/${runId}/history`
    );
  }
}
